File: Backend/Nodejs/src/models/artist.ts
import { formatDate } from "../utils/formatDate";
import { pool } from "./database/connection";
import { Artist } from "./types";

export class ArtistModel {
  static async createArtist(
    artist: Artist,
    photo: Express.MulterS3.File,
    callback: Function,
  ) {
    try {
      const newArtist: Artist = { ...artist };
      newArtist.avatar = photo.location;
      // Guardar en DB
      pool.query("CALL CreateArtist(?,?,?)", [
        newArtist.name,
        newArtist.avatar,
        formatDate(newArtist.birthDate),
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      callback(error, false);
    }
  }

  static getArtist({ id }: { id: number }, callback: Function) {
    try {
      pool.query(
        `SELECT a.id_artist AS id, a.name,a.image AS cover,a.birthdate  FROM Artists a WHERE a.id_artist=${id}`,
        (error, result, fields) => {
          if (error) throw error;
          callback(result[0], true);
        },
      );
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static getArtists(callback: Function) {
    try {
      pool.query(
        `SELECT a.id_artist AS id, a.name,a.image AS cover,a.birthdate  FROM Artists a`,
        (error, result) => {
          if (error) throw error;
          callback(result, true);
        },
      );
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static getSongs({ id }: { id: number }, callback: Function) {
    try {
      pool.query("CALL GetArtistSongs(?)", [
        id,
      ], (err, result) => {
        if (err) throw err;
        callback(result[0], true);
      });
    } catch (error) {
      callback(error, false);
    }
  }

  static getSongsNotInAlbum({ id }: { id: number }, callback: Function) {
    try {
      pool.query("CALL GetArtistSongsNotInAlbum(?)", [
        id,
      ], (err, result) => {
        if (err) throw err;
        callback(result[0], true);
      });
    } catch (error) {
      callback(error, false);
    }
  }

  static editArtist({ data }: { data: any }) {
  }

  static deleteArtist({ id }: { id: number }, callback: Function) {
    try {
      pool.query("CALL DeleteArtist(?)", [
        id,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      callback(error, false);
    }
  }
}

File: Backend/Python/src/models/artist.py
from src.models.database.connection import getCnx
from src.utils.formatDate import formatDate


class ArtistModel:

    @staticmethod
    def create_artist(artist, avatar_filename):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            birth_date = formatDate(artist["birthDate"])
            # Query
            cursor.callproc(
                "CreateArtist",
                (
                    artist["name"],
                    avatar_filename,
                    birth_date,
                ),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    return result, False
                else:
                    return result, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def get_artist(id):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Ejecuta la consulta SQL para obtener el usuario por su email
            cursor.execute(
                f"SELECT a.id_artist AS id, a.name,a.image AS cover,a.birthdate  FROM Artists a WHERE a.id_artist='{id}'"
            )
            result = cursor.fetchone()
            if result is not None and len(result) > 0:
                # Convierte el resultado en un diccionario y lo devuelve
                response = dict(zip(cursor.column_names, result))
                return response, True
            else:
                return {"MESSAGE": "Artista no encontrado", "TYPE": "ERROR"}, False
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()


    @staticmethod
    def get_artists():
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "GetAllArtists",
            )
            data = []
            for result in cursor.stored_results():
                for song in result.fetchall():
                    result = dict(zip(("id", "name", "cover"), song))
                    data.append(result)
            return data, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()


    @staticmethod
    def get_songs(artist):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "GetArtistSongs",
                (artist,),
            )
            data = []
            for result in cursor.stored_results():
                for song in result.fetchall():
                    result = dict(zip(("id", "name", "cover", "musicSrc"), song))
                    data.append(result)
            return data, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def edit_artist(artist, avatar_filename):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            birth_date = formatDate(artist["birthDate"])
            # crypto
            # Query
            cursor.callproc(
                "UpdateArtist",
                (
                    artist["id"],
                    artist["name"],
                    avatar_filename,
                    birth_date,
                ),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    return result, False
                else:
                    return result, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def delete_artist(id):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "DeleteArtist",
                (id,),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    return result, False
                else:
                    return result, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()

    @staticmethod
    def song_not_album(id):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc("GetArtistSongsNotInAlbum", (id,))
            data = []
            for result in cursor.stored_results():
                for song in result.fetchall():
                    result = dict(
                        zip(("id", "name", "cover", "musicSrc", "album"), song)
                    )
                    data.append(result)
            return data, True
        except Exception as e:
            return str(e), False
        finally:
            cursor.close()
